File: python/Problem004.py
#
# This file contains a solution to the problem number 4 of the Project Euler, last modified on 5/3/2018.
# Copyright (c) Matthieu Le Boucher. All rights reserved.
#

def result():
    maximalProduct = 0

    for i in range(999, 100, -1):
        for j in range(i, 100, -1):
            product = i * j

            if product > maximalProduct:
                productString = str(product)

                if productString == productString[::-1]:
                    maximalProduct = product

    return maximalProduct

if __name__ == "__main__":
    print(result())

File: cpp/Problem004.cpp
#include "EulerSolution.h"
#include <string>
#include <algorithm>

/*
 * @author Matthieu Le Boucher <matt.leboucher@gmail.com>
 */

using namespace std;

class Problem004 : public EulerSolution {
public:
    int number() override {
        return 4;
    }

    double solution() override {
        int maximalProduct = 0;
        int product;
        string productString;

        for (int i = 999; i >= 100; i--) {
            for (int j = i; j >= 100; j--) {
                product = i * j;

                if (product > maximalProduct) {
                    productString = to_string(product);
                    string reversedString(productString);
                    reverse(reversedString.begin(), reversedString.end());

                    if (productString == reversedString)
                        maximalProduct = product;
                }
            }
        }

        return maximalProduct;
    }

    double answer() override {
        return 906609;
    }
};